Optimise Everything

cristicbz opened the pull request:
Performance is all that matters, Rust is a systems programming language.

This commit implements the following optimisations:
1. Avoid using the `format!` machinery (via `println!` in a tight loop).
2. Avoid allocating a `Vec` of arguments when we just need the second one.
3. Avoid locking `stdout` over and over again in a tight loop, by hoisting
   `stdout.lock()` outside the loop.
4. Use buffered output.

Points 1, 3 and 4 improve overall performance, while point 2 improves start up speed. **Please review carefully.**

This commit increases output throughput as measured by (_short_)

```
target/release/yes | pv > /dev/null
```

and (_long_)

```
target/release/yes aaaaaaaaaaaaaaaa | pv > /dev/null
```

after 20s from:

| Implementation | Throughput (_short_) | Throughput (_long_) |
| --- | --- | --- |
| Old | 3.8 MB/s | 30.2 MB/s |
| New | **4.5 GB/s** | **4.5 GB/s** |
| GNU coreutils 8.21 | 140MB/s | 769MB/s |

This commit also improves error handling by breaking out of the loop on an I/O error.

This PR is also ridiculous.

**EDIT**

I added two new commits with the following additional optimisations:
    1. Do not allocate in the `"y\n"` case by using `Cow` (suggested by @eddyb and @withoutboats)
    2. Use `memcpy`-doubling to write progressively bigger buffers. Starts with a single copy of the output, then two, three etc. up until 64KB (suggested by @Veedrac).

Point (2) in particular increased throughput by an order of magnitude and it has the added advantage that time to first byte is really low compared to a fully-buffered approach.

Thanks everyone for contributing. It is very important to get this sort of thing right or people will never migrate off of C++.

**EDIT the second**

Now we also support non-unicode arguments on Unix!

cristicbz committed:
Optimise everything.

Performance is everything, Rust is a systems programming language.

This commit implements the following optimisations:
  1. Avoid using the `format!` machinery (via `println!` in a tight loop).
  2. Avoid allocating a `Vec` of arguments when we just need the second one.
  3. Avoid locking `stdout` over and over again in a tight loop, by hoisting
    `stdout.lock()` outside the loop.
  4. Fills up a buffer using `memcpy`-doubling and outputs that.
  5. Use `OsStr` to support non-unicode arguments on Unix.

Points 1, 3 and 4 improve overall performance, while point 2 improve start up
speed.

This commit increases output throughput as measured by

    target/release/yes | pv > /dev/null

and

    target/release/yes aaaaaaaaaaaaaaaa | pv > /dev/null

after 20s from (3.8 MB/s, 30.2 MB/s) to 4.5GB/s. As a baseline,
the GNU coreutils 8.21 implementation gives (140MB/s, 769MB/s).

This commit also improves error handling by breaking out of the loop on an I/O
error.

This commit is also ridiculous.

Many thanks to @eddyb, @Veedrac, @withoutboats, @stebalien, @llogiq for all
their suggestions.

eddyb commented on src/main.rs:
```diff
@@ -1,10 +1,21 @@
+            .skip(1)
+            .next()
+            .map(|x| x + "\n")
+            .unwrap_or("y\n".to_owned());
```
Allocation can be avoided here if not needed, with `.unwrap_or_else(|| "y\n".to_owned())`.
However, allocating is unnecessary in all cases, you could have one variable for a `String` and another one for `&str`, sometimes borrowing into the `String` variable.

withoutboats replied:
Rather than have two variables, output could be a `Cow<'static, str>` with this code:

```
        let output = env::args()
            .skip(1)
            .next()
            .map(|x| Cow::Owned(x + "\n"))
            .unwrap_or(Cow::Borrowed("y\n"));
```

cristicbz replied:
Thanks @eddyb and @withoutboats, fixed in 3552ef9.

Stebalien commented on src/main.rs:
```diff
@@ -1,10 +1,21 @@
+            .unwrap_or("y\n".to_owned());
+        let stdout = io::stdout();
+        let mut buffered = BufWriter::new(stdout.lock());
+        loop {
```
FYI, you can write this loop as:

``` rust
while let Ok(_) = buffered.write(output.as_bytes()) {}
```

frewsxcv replied:
Or even:

``` rust
while buffered.write(&output).is_ok() {}
```

Or maybe even:

``` rust
buffered.write_all(&output);
```

Stebalien replied:
Actually, it should probably be:

``` rust
while buffered.write_all(&output).is_ok() {}
```

(he's trying to repeatedly write the same string).

nathanross commented:
hey @cgati and @cristicbz , would you be willing to license your code on this using the MIT license in an MIT-compatible way like public domain? This is definitely higher throughput (having just tested) than the current version in [uutils/coreutils](https://github.com/uutils/coreutils/blob/a2944e06db93c70f20556fdd5039abfabce78be6/src/yes/yes.rs)

Mark-Simulacrum commented:
I get speeds of 10.4 GB/s with `yes` and 10.2 GB/s with `yes yyyyyyyyyyyyyyyyyyyy` with the following code. GNU coreutils v8.25 has exactly the same speed with the `yes` and 10.4 GB/s with the long invocation. I'm not sure why it wins with the long invocation.

With the current code in this PR, I get ~790 MB/s with the short invocation and ~1.1 GB/s with the long invocation. The problem here seems to lie in that BufWriter's `write` implementation will be fastest when writing the bytes immediately; I've removed buffering entirely to get the speeds above. It's possible the startup time could be further optimized through reducing allocations.

``` rust
use std::env;
use std::io::{self, Write};
use std::process;

const PER_WRITE: usize = 8 * 1024;

fn main() {
    {
        // Move all locals into a scope, so that `process::exit()` is called after all destructors.
        let output = env::args().nth(1)
            .map(|x| x + "\n")
            .unwrap_or("y\n".to_owned());
        let stdout = io::stdout();
        let mut locked = stdout.lock();

        let len = PER_WRITE / output.len();
        let buffer = std::iter::repeat(output).take(len).collect::<String>();

        loop {
            if locked.write(buffer.as_bytes()).is_err() {
                break;
            }
        }
    }
    process::exit(1);
}
```

Stebalien commented:
FYI, constructing the buffer that way will reallocate the string multiple times (collect has no way of knowing how long each of the `output` strings is). Instead, I recommend:

``` rust
let mut buffer = String::with_capacity(PER_WRITE);
buffer.extend(std::iter::repeat(output).take(len));
```

(although the syscalls will dominate anyways so this is probably a useless optimization...)

Mark-Simulacrum commented:
That may have helped with startup time, difficult to check...
However, I did just check another thing: `while _.is_ok() {}` instead of the loop/break in the code above is slower.

serprex commented:
If the string is smaller than 4kb then it should be repeated to that size. Then print with http://mainisusuallyafunction.blogspot.ca/2014/09/raw-system-calls-for-rust.html (syscalls crate now supports more than x86_64 linux)

eddyb commented:
@serpex Unbuffered stdout calls `write` which ends up in the syscall. There should be no overhead other than the calls themselves.

cgati commented:
@nathanross absolutely. I'm currently out of town without my laptop for my sister's graduation, but as soon as I'm home this evening I'd be happy to.

Veedrac commented:
@Mark-Simulacrum That code fails if the input string is longer than the buffer. You should do

```
let len = PER_WRITE / output.len() + 1;
```

to avoid that. Further, all of these break if `locked.write` doesn't write all of the bytes - they should be using `write_all`. That can have a performance overhead (I dunno, I never checked), so really you want:

```
let len = std::cmp::max(PER_WRITE, output.len());
let buffer: Vec<u8> = output.bytes().cycle().take(len).collect();

let mut offset = 0;
while let Ok(bytes) = locked.write(&buffer[offset..]) {
    offset += bytes;
    offset %= output.len();
}
```

You probably also want to use `args_os`, since `yes` doesn't require unicode. I don't know a pleasant platform-independant way to turn an `OsStr` to a `&[u8]`, so you'll need platform-specific code. It just so turns out that `yes` actually takes multiple arguments, too, separating them with spaces. And a `--help` argument, and a `--version` argument... we should support those. And the `yes -- foo` argument syntax. You know, I'm really starting to doubt my sanity. But that's fine... let's let's optimize the startup time too. that would be a Good Idea. How about we dispatch off of the `env::args` and IF it's length one then IF the argument is longer than `PER_WRITE` we just convert it to a `Vec<u8>` directly (except on Windo`ws) b`ecause that avoids generating the buffer but we should Also specialize on zer0 length because we can generate that with specialized code `buffer = vec![b'\n'; PER_WRITE]; for i in 0..PER_WRITE / 2 { buffer[i * 2] = b'y'; }` but then for other lengths we can use m`emcpy`-based doubling like while i < buffer.len() { let (lhs, mut rhs) = buffer.split_at_mut(i); let _ = rhs.write(lhs); i *= 2; } then it's fast.

cristicbz commented:
Excellent, thanks a lot @eddyb, @withoutboats, @Stebalien and @Veedrac , I implemented your suggestions, for a grand total of 4.5 GB/s throughput! See 08116fb.

cristicbz commented:
Hey @nathanross , despite the extreme innovation displayed in this PR, I will generously dual licence my contributions as MIT/Apache-2.0. Though I imagine that for uutils, you want more features like `--version` etc, right?

nathanross commented:
@cristicbz @cgati great! Thank you.

@cristicbz yes, that the uutils `yes` will support that flag; I will build the commit next weekend if no one else has by then.

aij commented on src/main.rs:
```diff
@@ -1,23 +1,35 @@
+    let _ = (&mut buffer[..]).write(output_bytes);
+    while buffer_size < BUFFER_CAPACITY {
+        let (left, mut right) = buffer.split_at_mut(buffer_size);
+        if !locked.write_all(left).is_ok() {
```
Did you benchmark this? How much does it affect startup time? Or is it a premature optimization?

aij commented on src/main.rs:
```diff
@@ -1,23 +1,35 @@
+    let mut buffer_size = output_bytes.len();
+
+    let _ = (&mut buffer[..]).write(output_bytes);
+    while buffer_size < BUFFER_CAPACITY {
```
Wouldn't this break when output_bytes is not a power of 2?

mmstick commented on src/main.rs:
```diff
@@ -1,10 +1,35 @@
-    match arguments.len() {
-        1 => loop { println!("y") },
-        _ => loop { println!("{}", arguments[1]) },
+const BUFFER_CAPACITY: usize = 64 * 1024;
```
Windows can only handle buffers of 16K size for standard output/input, versus 64K sizes used by everyone else.

llogiq commented on src/main.rs:
```diff
@@ -1,10 +1,35 @@
+fn main() {
+    write(env::args()
+        .skip(1)
+        .next()
```
`.nth(1)`

cristicbz replied:
Good point. Is there a clippy lint about this? Did you run clippy @llogiq? Knowing you, you probably did, despite the 35LOC size...

llogiq replied:
Yes, there is a clippy lint. And no, I was on mobile, so I didn't run it – I guess I sort of internalized it.

llogiq commented on src/main.rs:
```diff
@@ -1,10 +1,35 @@
+        .skip(1)
+        .next()
+        .map(|x| Cow::Owned(x + "\n"))
+        .unwrap_or(Cow::Borrowed("y\n"))
```
`.map_or(..)`

cristicbz replied:
Ugh, I hate the argument order on `map_or`so much that I block the memory of its existence. Anyway. Will fix.

cristicbz commented:
I implemented all your suggestions, including `OsString` to support non-unicode strings (though on non-unix, I still roundtrip through UTF-8; we can leave this for another PR). OS specific support is encapsulated in the `platform` module.

Also fixed the edge cases around the buffer size vs output size. 

cc @cgati @llogiq @Stebalien @Stebalien @eddyb @Veedrac 

Please have one last look, let's all pull together and land this momentous PR.

aij commented on src/main.rs:
```diff
@@ -1,10 +1,64 @@
+
+    while buffer_size < buffer.len() {
+        let (left, mut right) = buffer.split_at_mut(buffer_size);
+        right.write(left).unwrap();
```
What does this do when `left` is longer than `right`? It's hard to imagine it would stop on the last `\n` that fits...

cristicbz replied:
Hmm, I did try this in the terminal, but maybe didn't manage to trigger this, I think you're right that this should probably check `< buffer.len() / 2`

aij replied:
Ideally, I would think you would want to fill up as much of the buffer as possible, to maximize the data written with each syscall.

Either way, you have to be careful to write() only the data that was initialized, or the output would be incorrect.

Wouldn't this all be simpler if you used a pre-allocated `Vec`, or does that add inefficiency?

aij commented on src/main.rs:
```diff
@@ -1,10 +1,64 @@
+    let mut locked = stdout.lock();
+    let mut buffer = [0u8; BUFFER_CAPACITY];
+
+    while locked.write_all(fill_up_buffer(&mut buffer, output)).is_ok() {}
```
Isn't this going to rewrite the buffer each time?

cristicbz replied:
Oh wow, excellent point. Throughput isn't affected though, which makes me thing either filing it up is stupid fast, or llvm hoists it out of the loop. Will fix that. I've been really careless with this pr...

aij replied:
If LLVM already does optimize it, I'm not sure there's much reason to do so manually. I might look at it this evening, but I don't have access to a rustc right now.

cgati commented:
LGTM. I'll go ahead and merge this in. Thank you everyone for your suggestions and all the work done here!

mmstick commented:
The GNU `yes` command actually accepts multiple arguments, instead of just the first argument.

cgati commented:
@mmstick I originally wrote this to mimic the BSD `yes` as found on macOS (which is non-obvious, as I never stated that before.) That could probably be filed as an issue, however.

aij commented:
@mmstick I was wondering whether that was standard or a GNU extension.

GNU yes also tries to parse options, which I believe is nonstandard. Eg:

```
$ yes --verbose
yes: invalid option -- '-'
Try 'yes --help' for more information.
```

sugar700 commented:
`yes` is not POSIX, so it's hard to say what are extensions.
